[DeviceMesh] Make CuTe layout as mesh layout to be ready for using in DeviceMesh

fduwjj opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #161106
* #161016
* #162690
* __->__ #162414
* #162534
* #162413

We create a wrapper class named "_MeshLayout" acting as a layout for device mesh so that we can add new methods more specific to DeviceMesh and keep the core logic of CuTe manipulation inside pycute module. This PR create the main body of the code and then next PR will come with actual implementation and unit test for device mesh layout. (Actual implementation can be found in https://github.com/pytorch/pytorch/pull/161016)

cc @H-Huang @awgu @wanchaol @fegin @wz337 @wconstab @d4l3k @pragupta @ezyang @msaroufim @dcci

fduwjj committed:
Make CuTe layout as mesh layout to be ready for using in DeviceMesh

[ghstack-poisoned]

fduwjj committed:
Update on "Make CuTe layout as mesh layout to be ready for using in DeviceMesh"




cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k pragupta ezyang msaroufim

[ghstack-poisoned]

ezyang commented:
ngl, the size of this diff is NOT looking good. What's going on?

fduwjj commented:
@ezyang This one is still WIP and I will need to update its descriptions. Basically I want to do two things in this PR: To make API into layout class instead of using original pycute implementation. So it is purely cosmetic change and remove some part of functions which are not used by us, but still no change to the business logic.

fduwjj committed:
Update on "Make CuTe layout as mesh layout to be ready for using in DeviceMesh"




cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k pragupta ezyang msaroufim

[ghstack-poisoned]

fduwjj committed:
Update on "Make CuTe layout as mesh layout to be ready for using in DeviceMesh"




cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k pragupta ezyang msaroufim

[ghstack-poisoned]

fduwjj committed:
Update on "Make CuTe layout as mesh layout to be ready for using in DeviceMesh"




cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k pragupta ezyang msaroufim

[ghstack-poisoned]

fduwjj commented:
@ezyang yes they already have that Layout class but all methods are not members of that class unfortunately we need to move it into the class.

fduwjj committed:
Update on "Make CuTe layout as mesh layout to be ready for using in DeviceMesh"


Original pycute code does not include manipulation methods as the member methods of layout class, so this PR did that and add type checker, we don't really change the logic of functions and most material change is in the next PR of this stack so that code review will be easier.


cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k pragupta ezyang msaroufim dcci

[ghstack-poisoned]

fduwjj committed:
Update on "Make CuTe layout as mesh layout to be ready for using in DeviceMesh"


We create a wrapper class acting as a layout for device mesh so that we can modify existing implementation of PyCuTe layout without modifying original code. This PR create the main body of the code and then next PR will come with actual implementation and unit test for device mesh layout.

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k pragupta ezyang msaroufim dcci

[ghstack-poisoned]

fduwjj committed:
Update on "[DeviceMesh] Make CuTe layout as mesh layout to be ready for using in DeviceMesh"


We create a wrapper class acting as a layout for device mesh so that we can modify existing implementation of PyCuTe layout without modifying original code. This PR create the main body of the code and then next PR will come with actual implementation and unit test for device mesh layout.

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k pragupta ezyang msaroufim dcci

[ghstack-poisoned]

fduwjj commented:
As discussed offline, split the PR so that it's easier to review as well.

ezyang commented on torch/distributed/_mesh_layout.py:
```diff
@@ -0,0 +1,75 @@
+
+    @property
+    def sizes_and_strides(self) -> Iterator[tuple[int, int]]:
+        return zip(flatten(self.shape), flatten(self.stride))  # type: ignore[arg-type]
```
OK, it's a lot clearer in this formulation of the PR. This method makes me nervous; it's not obvious to me that downstream people will use this correctly.

ezyang commented on torch/distributed/_mesh_layout.py:
```diff
@@ -0,0 +1,75 @@
+        elif isinstance(size, int) and isinstance(stride, int):
+            return _Layout((size,), (stride,))
+        else:
+            raise ValueError("size and stride must be either int or tuple")
```
Why doesn't this directly use Layout.__getitem__

fduwjj commented on torch/distributed/_mesh_layout.py:
```diff
@@ -0,0 +1,75 @@
+
+
+@dataclass(frozen=True, init=True)
+class _Layout(Layout):
```
so ok that is something I want to discuss. The original pycute assume the size and stride can be just a int which makes their code very complicated so I really don't like it... Plus we will add more device-mesh bookkeeping related logics into this class as well, so that's why I want to have this wrapper class.

fduwjj committed:
Update on "[DeviceMesh] Make CuTe layout as mesh layout to be ready for using in DeviceMesh"


We create a wrapper class acting as a layout for device mesh so that we can modify existing implementation of PyCuTe layout without modifying original code. This PR create the main body of the code and then next PR will come with actual implementation and unit test for device mesh layout.

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k pragupta ezyang msaroufim dcci

[ghstack-poisoned]

fduwjj committed:
Update on "[DeviceMesh] Make CuTe layout as mesh layout to be ready for using in DeviceMesh"


We create a wrapper class acting as a layout for device mesh so that we can modify existing implementation of PyCuTe layout without modifying original code. This PR create the main body of the code and then next PR will come with actual implementation and unit test for device mesh layout.

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k pragupta ezyang msaroufim dcci

[ghstack-poisoned]

fduwjj committed:
Update on "[DeviceMesh] Make CuTe layout as mesh layout to be ready for using in DeviceMesh"


We create a wrapper class acting as a layout for device mesh so that we can modify existing implementation of PyCuTe layout without modifying original code. This PR create the main body of the code and then next PR will come with actual implementation and unit test for device mesh layout.

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k pragupta ezyang msaroufim dcci

[ghstack-poisoned]

fduwjj committed:
Update on "[DeviceMesh] Make CuTe layout as mesh layout to be ready for using in DeviceMesh"


We create a wrapper class acting as a layout for device mesh so that we can add new methods more specific to DeviceMesh and keep the core logic of CuTe manipulation inside pycute module. This PR create the main body of the code and then next PR will come with actual implementation and unit test for device mesh layout. (Actual implementation can be found in https://github.com/pytorch/pytorch/pull/161016)

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k pragupta ezyang msaroufim dcci

[ghstack-poisoned]

fduwjj committed:
Update on "[DeviceMesh] Make CuTe layout as mesh layout to be ready for using in DeviceMesh"


We create a wrapper class acting as a layout for device mesh so that we can add new methods more specific to DeviceMesh and keep the core logic of CuTe manipulation inside pycute module. This PR create the main body of the code and then next PR will come with actual implementation and unit test for device mesh layout. (Actual implementation can be found in https://github.com/pytorch/pytorch/pull/161016)

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k pragupta ezyang msaroufim dcci

[ghstack-poisoned]

ezyang reviewed (approved):
Nothing here seems objectionable

fduwjj committed:
Update on "[DeviceMesh] Make CuTe layout as mesh layout to be ready for using in DeviceMesh"


We create a wrapper class acting as a layout for device mesh so that we can add new methods more specific to DeviceMesh and keep the core logic of CuTe manipulation inside pycute module. This PR create the main body of the code and then next PR will come with actual implementation and unit test for device mesh layout. (Actual implementation can be found in https://github.com/pytorch/pytorch/pull/161016)

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k pragupta ezyang msaroufim dcci

[ghstack-poisoned]

fduwjj committed:
Update on "[DeviceMesh] Make CuTe layout as mesh layout to be ready for using in DeviceMesh"


We create a wrapper class named "_MeshLayout" acting as a layout for device mesh so that we can add new methods more specific to DeviceMesh and keep the core logic of CuTe manipulation inside pycute module. This PR create the main body of the code and then next PR will come with actual implementation and unit test for device mesh layout. (Actual implementation can be found in https://github.com/pytorch/pytorch/pull/161016)

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k pragupta ezyang msaroufim dcci

[ghstack-poisoned]

fduwjj commented:
@pytorchbot merge
